File: objc/Visitor/Visitor/AveragePriceCalcVisitor.m
#import "AveragePriceCalcVisitor.h"

@implementation AveragePriceCalcVisitor {
    double totalPrice;
    double avgPrice;
}

-(instancetype)init {
    if(self = [super init]) {
        totalPrice = avgPrice = 0.0;
    }
    return self;
}

-(double)getAvgPrice {
    return avgPrice;
}

-(void)visitBook:(Book*)book {
    totalPrice += book.price;
}

-(void)visitLibrary:(Library*)library {
    avgPrice = totalPrice / library.bookList.count;
}

@end

File: swift/Visitor/Visitor/AveragePriceCalcVisitor.swift
import Foundation

class AveragePriceCalcVisitor : LibraryVisitor {
    var totalPrice: Double = 0.0
    var avgPrice: Double = 0.0
    
    func getAvgPrice() -> Double {
        return avgPrice
    }
    
    func visitBook(_ book: Book) {
        totalPrice += Double(book.price)
    }
    
    func visitLibrary(_ library: Library) {
        avgPrice = totalPrice / Double(library.bookList.count)
    }
}
